runtime/src/main/java/org/corfudb/runtime/view/ObjectsView.java

```diff
@@ -157,6 +157,12 @@ public long TXEnd() throws TransactionAbortedException {
             long commitAddress = TransactionalContext.getCurrentContext().commitTransaction();
             MicroMeterUtils.time(Duration.ofMillis(System.currentTimeMillis() - context.getStartTime()),
                     "transaction.duration");
+            if (commitAddress == Address.NON_ADDRESS) {
+                // If no streams were touched or only empty streams were touched, the returned address would be -1
+                // But -1 can be detrimental to stream subscription which is just looking for a safe spot
+                // to resume subscription from, so instead return the address from the snapshot token taken.
+                return context.getSnapshotTimestamp().getSequence();
+            }
```

Comment:
In the case of an empty transaction, or empty stream reads which will return the global tail which is not materialized and thus not reliable to use with the streaming subscribe methods. 

Since this is not enough, the shim layer should make sure to materialize the snapshot it before it is used. The easiest way to do that is to read the snapshot returned by the transaction.

Reply:
Sure, I can the add a read, but wondering if that can be expensive?
Let's say the streaming layer is given an un-materialized global tail, won't it be able to handle the different cases associated with it like a sequencer regression or slow writer since these cases can happen normally as well?

Reply:
@hisundar if I understand it right, @Maithem is talking about a functional/correctness issue but you are talking about performance hit with the read. Do you have any other way not have that read and yet solve the problem?

Reply:
Just bringing this up as it is also needed by LR (we just had a discussion on this, this week). I think it's a good idea to either:
(a) Provide an API through the Log Unit that retrieves the actual materialized tail or;
(b) Provide/Expose a function materializeTailAndGet(), which does what Maithem describes, enforces a read on the tail, to guarantee that the log is materialized up to that point. 

This function would be used in any case where potential sequencer regressions can lead to data loss. LR and this case are some examples that could make use of this. I agree there is a cost for a single read, but that's a cost we need to pay in order to avoid bigger issues.

Reply:
Performance vs Correctness is not a trade-off in this context: consistency is a requirement here.

This semantic is very specific to the UFO Table/Stream abstractions so this change should be moved so that it is confined to UFO only. 

As for the cost of the extra read, it shouldn't be expensive because it will only be incurred in the rare case where the Table is not checkpointed, that is when the system is first bootstrapped, or when the transaction is empty during a full sync, which is usually not the case.

In essence, the extra cost should only be incurred in a rare edge case to maintain correctness and won't penalize transactions otherwise.

Reply:
Agreed, added the read. thanks